src/lib/util/simh_tape_file.cpp

```diff
@@ -0,0 +1,624 @@
+// license:BSD-3-Clause
+// copyright-holders:Mietek Bak
+
+// best read together with SIMH magtape spec (rev 17 Jan 2022)
+// http://simh.trailing-edge.com/docs/simh_magtape.pdf
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include "multibyte.h"
+#include "simh_tape_file.h"
+
+#if 0
+#define LOG(...)    fprintf(stderr, __VA_ARGS__)
+#else
+#define LOG(...)    ([](auto &&... args){})(__VA_ARGS__)
+#endif
+
+#define SIMH_TAPE_MARK                      0x00000000
+#define SIMH_ERASE_GAP                      0xfffffffe
+#define SIMH_HALF_GAP_FORWARD               0xfffeffff
+#define SIMH_END_OF_MEDIUM                  0xffffffff
+
+#define SIMH_GOOD_DATA_RECORD               0x0
+#define SIMH_PRIVATE_DATA_RECORD_1          0x1
+#define SIMH_PRIVATE_DATA_RECORD_2          0x2
+#define SIMH_PRIVATE_DATA_RECORD_3          0x3
+#define SIMH_PRIVATE_DATA_RECORD_4          0x4
+#define SIMH_PRIVATE_DATA_RECORD_5          0x5
+#define SIMH_PRIVATE_DATA_RECORD_6          0x6
+#define SIMH_PRIVATE_MARKER                 0x7
+#define SIMH_BAD_DATA_RECORD                0x8
+#define SIMH_RESERVED_DATA_RECORD_9         0x9
+#define SIMH_RESERVED_DATA_RECORD_A         0xa
+#define SIMH_RESERVED_DATA_RECORD_B         0xb
+#define SIMH_RESERVED_DATA_RECORD_C         0xc
+#define SIMH_RESERVED_DATA_RECORD_D         0xd
+#define SIMH_TAPE_DESCRIPTION_DATA_RECORD   0xe
+#define SIMH_RESERVED_MARKER                0xf
+
+inline const int get_marker_class(const int marker) { return marker >> 28; }
+inline const int get_marker_value(const int marker) { return marker & 0x0fffffff; }
+inline const bool is_marker_half_gap_reverse(const int marker) { return marker >= 0xffff0000 && marker <= 0xfffffffd; }
+
+//////////////////////////////////////////////////////////////////////////////
+
+// construction
+
+simh_tape_file::simh_tape_file(const std::string_view path, const bool create)
```

Comment:
I think being able to read/write SIMH format is perfectly reasonable, but I'm not sure it's the best candidate for a canonical MAME tape format; I don't have any suggestions other than perhaps reviewing whether CHD might be able to be enhanced for the purpose? One significant gap in my mind is that the SIMH format does not appear to have any way to represent what kind of tape or tape drive the image corresponds to. It should not be possible to insert a DAT tape in a QIC drive, for example.

Reply:
I don't think SIMH needs to be declared canonical.  If you have something else in mind, go for it.  But interoperating with SIMH is definitely nice as there are hundreds of tapes out there in its format.

Reply:
I’ll make it clear that the SIMH format is one of potentially many supported formats by introducing an abstract tape file interface.
```c++
enum class tape_status : int {
	OK,
	BOM, // beginning of medium
	EW, // early warning
	FILEMARK,
	FILEMARK_EW, // filemark and early warning
	SETMARK,
	SETMARK_EW, // setmark and early warning
	EOD, // end of data
	EOD_EW, // end of data and early warning
	UNKNOWN,
	UNKNOWN_EW, // unknown and early warning
	EOM // end of medium
};

class tape_file_interface
{
public:
	// construction
	virtual ~tape_file_interface() {}

	// position-preserving operations
	virtual const bool is_read_only() const = 0;
	virtual const bool is_ew() const = 0;
	virtual const uint8_t get_density_code() const = 0;
	virtual const std::pair<const tape_status, const int> read_position() const = 0;

	// non-destructive operations
	virtual void rewind(const bool eom) = 0;
	virtual const tape_status locate_block(const int req_block_addr) = 0;
	virtual const tape_status space_eod() = 0;
	virtual const std::pair<const tape_status, const int> space_blocks(const int req_blocks_num) = 0;
	virtual const std::pair<const tape_status, const int> space_filemarks(const int req_marks_num, const bool setmarks = false, const bool sequential = false) = 0;
	virtual const std::pair<const tape_status, const int> space_blocks_reverse(const int req_blocks_num) = 0;
	virtual const std::pair<const tape_status, const int> space_filemarks_reverse(const int req_marks_num, const bool setmarks = false, const bool sequential = false) = 0;

	virtual const std::pair<const tape_status, const int> read_block(uint8_t *buf, const int buf_size) = 0;

	// destructive operations
	virtual void erase(const bool eom) = 0;
	virtual const tape_status write_block(const uint8_t *const buf, const int len) = 0;
	virtual const tape_status write_filemarks(const int req_marks_num, const bool setmarks = false) = 0;
};
```